src/main/ncdio_pio.F90.in

```diff
@@ -1902,18 +1940,40 @@ contains
              if (present(nt)) then
                 call pio_setframe(ncid, vardesc, int(nt,kind=Pio_Offset_Kind))
              end if
+#if ({ITYPE}==TYPEINT)
+             if(xtype == PIO_INT .or. xtype == PIO_REAL) then
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, data, status)
+             elseif(xtype == PIO_DOUBLE) then
+                allocate(ddata(size(data,1),size(data,2)))
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, ddata, status)
+                data = int(ddata)
+             endif
+#endif
+#if ({ITYPE}==TYPEDOUBLE)
+             if(xtype == PIO_DOUBLE) then
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, data, status)
+             else if(xtype == PIO_INT) then
+                allocate(idata(size(data,1),size(data,2)))
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, idata, status)
+                data = real(idata, kind=R8)
+                deallocate(idata)
+             else if(xtype == PIO_REAL) then
+                allocate(fdata(size(data,1),size(data,2)))
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, fdata, status)
+                data = real(fdata, kind=R8)
+                deallocate(fdata)
+             endif
+#endif
```

Comment:
- [ ] I'm having trouble seeing how this block of code works as it currently is if switchdim is given: The code that was in place before called `pio_read_darray` to read into `temp`, which is the transpose of `data`. So it feels like it shouldn't work to read directly into data, as is now being done. Is there something that I'm not understanding that allows this to work as it is now?